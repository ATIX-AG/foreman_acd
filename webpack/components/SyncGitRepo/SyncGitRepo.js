import React, { useState } from 'react'
import PropTypes from 'prop-types';
import {
  Icon,
  Button,
} from 'patternfly-react';
import {
  sprintf,
  translate as __
} from 'foremanReact/common/I18n';
import {Select, TextInput} from 'foremanReact/components/common/forms/Select';
import RailsData from '../common/RailsData';
import ScmTypeSelector from './components/ScmTypeSelector';
import FormTextInput from './components/FormTextInput'
import { arrayToObject } from '../../helper';
import CommonForm from 'foremanReact/components/common/forms/CommonForm';
import $ from 'jquery';

import {
  FormControl,
  inlineEditFormatterFactory,
} from 'patternfly-react';

class SyncGitRepo extends React.Component {
  constructor(props) {
    super(props);
  }

  validateParameters() {
    let result = true;
    let msg = "";

    if (this.props.path == '' && this.props.scmType == 'directory') {
      result = false;

      if (msg == "") {
        msg += __("Directory path cannot be blank");
      }
    }

    if (this.props.gitUrl == '' && this.props.scmType == 'git') {
      result = false;

      if (msg == "") {
        msg += __("Git URL cannot be blank");
      }
    }

    return {
      validateResult: result,
      validateMsg: msg
    }
  }

  componentDidMount() {
    const {
      data: {mode, scmType, path, gitCommit, gitUrl, appDefinitions},
      initSyncGitRepo,
      loadScmType,
      loadPath,
      loadGitCommit,
      loadGitUrl,
      handleGitRepoSync,
    } = this.props;

    if (mode === 'editInstance') {
      loadScmType(scmType);
      loadPath(path);
      loadGitCommit(gitCommit);
      loadGitUrl(gitUrl);
      handleGitRepoSync(this.props.scmType, this.props.path, this.props.gitCommit)
    }

    initSyncGitRepo(
      scmType,
      path,
      gitCommit,
      gitUrl,
    );
  };

  render() {
    const {
      data: {mode, scmTypes, organization, location, appDefinitions},
      scmType,
      path,
      gitCommit,
      gitUrl,
      loadScmType,
      loadPath,
      loadGitCommit,
      loadGitUrl,
      handleGitRepoSync,
    } = this.props;

    var url_validator = /^(ftp|http|https):\/\/[^ "]+$/;

    let { validateResult, validateMsg } = this.validateParameters();

    if (validateResult == false) {
      $('input[type="submit"][name="commit"]').attr("disabled", true);
    } else {
      $('input[type="submit"][name="commit"]').attr("disabled", false);
    }

    return (
      <span>
        <div>
          <ScmTypeSelector
            label="SCM Type"
            editable={ appDefinitions.length == 0 }
            viewText={ scmTypes[scmType] }
            options={ scmTypes }
            onChange={ loadScmType }
            selectValue={ scmType }
          />
          {(scmType === 'directory') ? (
          <FormTextInput
            label= 'Directory Path *'
            editable= { appDefinitions.length == 0 }
            viewText={ path }
            onChange={ loadPath }
            parameter="path"
          /> ) : null }
          {(scmType === 'git') ? (
          <FormTextInput
            label= 'Git Url *'
            editable= { appDefinitions.length == 0 }
            viewText={ gitUrl }
            onChange={ loadGitUrl }
            parameter="git_url"
          /> ) : null }
          {(!url_validator.test(gitUrl) && gitUrl)? (
          <div className="form-group">
            <div className="col-md-2"></div>
            <div className="col-md-4"><pre>{ __("Git URL is wrong") }</pre></div>
          </div>
        ) : (<div></div>)}
        {validateResult == false ? (
            <div className="form-group">
              <div className="col-md-2"></div>
              <div className="col-md-4"><pre>{validateMsg}</pre></div>
            </div>
        ) : (<div></div>)}
          {(scmType === "git") ? (
          <FormTextInput
            label="Git Branch/Commit/Tag"
            editable= { appDefinitions.length == 0 }
            viewText={ gitCommit }
            onChange={ loadGitCommit }
            parameter="git_commit"
          /> ) : null }
          {(scmType === "git") ? (
          <CommonForm>
          <span className="help-block help-inline">
          <Button
            bsStyle="default"
            onClick={ (e) => handleGitRepoSync(gitUrl, gitCommit, scmType, e) }
            >
            {__('Sync Repository')} </Button>
            </span>
            </CommonForm>
        ) : null }

        </div>
          </span>
        )};
}

SyncGitRepo.defaultProps = {
  error: {},
  scmType: "",
  path: "",
  gitCommit: "",
  gitUrl: "",

}

SyncGitRepo.propTypes = {
  initSyncGitRepo: PropTypes.func,
  loadScmType: PropTypes.func,
  loadGitCommit: PropTypes.func,
  loadPath: PropTypes.func,
  loadGitUrl: PropTypes.func,
  handleGitRepoSync: PropTypes.func,
  scmType: PropTypes.string.isRequired,
  path: PropTypes.string.isRequired,
  gitCommit: PropTypes.string,
  gitUrl: PropTypes.string,
};

export default SyncGitRepo;
