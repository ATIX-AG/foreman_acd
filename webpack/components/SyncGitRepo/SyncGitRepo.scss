@import '~@theforeman/vendor/scss/variables';
