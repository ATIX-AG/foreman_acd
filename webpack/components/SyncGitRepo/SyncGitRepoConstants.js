export const SYNC_GIT_REPO_INIT = 'SYNC_GIT_REPO_INIT';
export const SYNC_GIT_REPO_LOAD_PATH = 'SYNC_GIT_REPO_LOAD_PATH';
export const SYNC_GIT_REPO_LOAD_SCM_TYPE = 'SYNC_GIT_REPO_LOAD_SCM_TYPE';
export const SYNC_GIT_REPO_LOAD_GIT_COMMIT = 'SYNC_GIT_REPO_LOAD_GIT_COMMIT';
export const SYNC_GIT_REPO_LOAD_GIT_URL = 'SYNC_GIT_REPO_LOAD_GIT_URL';
export const SYNC_GIT_REPO_REQUEST = 'SYNC_GIT_REPO_REQUEST';
export const SYNC_GIT_REPO_SUCCESS = 'SYNC_GIT_REPO_SUCCESS';
export const SYNC_GIT_REPO_FAILURE = 'SYNC_GIT_REPO_FAILURE';
export const SYNC_GIT_REPOSITORY = 'ANSIBLE_PLAYBOOK_SYNC_GIT_REPOSITORY';
