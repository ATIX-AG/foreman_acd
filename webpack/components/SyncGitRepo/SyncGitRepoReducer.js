import Immutable from 'seamless-immutable';

import {
  SYNC_GIT_REPO_INIT,
  SYNC_GIT_REPO_LOAD_SCM_TYPE,
  SYNC_GIT_REPO_LOAD_PATH,
  SYNC_GIT_REPO_LOAD_GIT_COMMIT,
  SYNC_GIT_REPO_LOAD_GIT_URL,
  SYNC_GIT_REPO_FAILURE,
  SYNC_GIT_REPO_REQUEST,
  SYNC_GIT_REPO_SUCCESS,
} from './SyncGitRepoConstants';

export const initialState = Immutable({
  loading: false,
});

const syncGitRepoConf = (state = initialState, action) => {
  const { payload } = action;

  switch (action.type) {
    case SYNC_GIT_REPO_INIT: {
      return state.merge(payload);
    }
    case SYNC_GIT_REPO_LOAD_SCM_TYPE: {
      let newState = {};
      const scmType = action.scmType;

      newState = {
        scmType: scmType,
      };

      return state.merge(newState);
    }
    case SYNC_GIT_REPO_LOAD_PATH: {
      let newState = {};
      const path = action.path;

      newState = {
        path: path,
      };

      return state.merge(newState);
    }
    case SYNC_GIT_REPO_LOAD_GIT_COMMIT: {
      let newState = {};
      const gitCommit = action.gitCommit;

      newState = {
        gitCommit: gitCommit,
      };

      return state.merge(newState);
    }
    case SYNC_GIT_REPO_LOAD_GIT_URL: {
      let newState = {};
      const gitUrl = action.gitUrl;

      newState = {
        gitUrl: gitUrl,
      };

      return state.merge(newState);
    }
    case SYNC_GIT_REPO_FAILURE: {
       return state.merge({ error: payload.error, loading: false });
     }
     case SYNC_GIT_REPO_REQUEST: {
       return state.set('loading', true);
     }
     case SYNC_GIT_REPO_SUCCESS: {
       let newState = {};
       return state.merge(newState);
     }
    default:
      return state;
  }
};

export default syncGitRepoConf;
