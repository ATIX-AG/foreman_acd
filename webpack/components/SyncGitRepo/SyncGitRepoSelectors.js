const syncGitRepoConf = state => state.foremanAcd.syncGitRepoConf;

export const selectScmType = state => syncGitRepoConf(state).scmType;
export const selectPath = state => syncGitRepoConf(state).path;
export const selectGitCommit = state => syncGitRepoConf(state).gitCommit;
export const selectGitUrl = state => syncGitRepoConf(state).gitUrl;
