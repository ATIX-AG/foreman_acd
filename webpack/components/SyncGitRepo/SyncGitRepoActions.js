import React from 'react';
import { API, actionTypeGenerator } from 'foremanReact/redux/API';
import { sprintf, translate as __ } from 'foremanReact/common/I18n';
import { addToast } from 'foremanReact/components/ToastsList';

import {
  SYNC_GIT_REPO_INIT,
  SYNC_GIT_REPO_LOAD_SCM_TYPE,
  SYNC_GIT_REPO_LOAD_PATH,
  SYNC_GIT_REPO_LOAD_GIT_COMMIT,
  SYNC_GIT_REPO_LOAD_GIT_URL,
  SYNC_GIT_REPO_REQUEST,
  SYNC_GIT_REPO_FAILURE,
  SYNC_GIT_REPO_SUCCESS,
} from './SyncGitRepoConstants';

export const initSyncGitRepo = (
  scmType,
  path,
  gitCommit,
  gitUrl
) => dispatch => {
  const initialState = {};

  initialState.scmType = scmType;
  initialState.path = path;
  initialState.gitCommit = gitCommit;
  initialState.gitUrl = gitUrl;

  dispatch({
    type: SYNC_GIT_REPO_INIT,
    payload: initialState,
  });
};

const errorHandler = (msg, err) => {
  const error = {
    errorMsg: __('Failed to fetch data from server.'),
    statusText: err,
  };
  return { type: msg, payload: { error } };
};

export const handleGitRepoSync = (
  gitUrl,
  gitCommit,
  scmType,
  e
) => async dispatch => {
  e.preventDefault();
  // const { REQUEST, SUCCESS, FAILURE } = actionTypeGenerator(SYNC_GIT_REPOSITORY);
  dispatch({
    type: SYNC_GIT_REPO_REQUEST,
    payload: {
      git_commit: gitCommit,
      scm_type: scmType,
      git_url: gitUrl,
    },
  });
  try {
    const { data } = await API.post('/acd/ansible_playbooks/sync_git_repo', {
      ansible_playbook: {
        git_commit: gitCommit,
        scm_type: scmType,
        git_url: gitUrl,
      },
    });
    dispatch(
      addToast({
        type: 'success',
        message: sprintf(__('Sucessfully synced git repository ')),
        key: SYNC_GIT_REPO_SUCCESS,
      })
    );
    return dispatch({
      type: SYNC_GIT_REPO_SUCCESS,
      payload: {
        git_commit: gitCommit,
        git_url: gitUrl,
        scm_type: scmType,
      },
      response: data,
    });
  } catch (error) {
    dispatch(
      addToast({
        type: 'error',
        message: sprintf(
          __('Error occurred while syncing git repository: %s'),
          error.response.data.message
        ),
        key: SYNC_GIT_REPO_FAILURE,
      })
    );
    return dispatch({
      type: SYNC_GIT_REPO_FAILURE,
      payload: {
        error,
      },
      response: error,
    });
  }
};

export const loadScmType = scmType => ({
  type: SYNC_GIT_REPO_LOAD_SCM_TYPE,
  scmType,
});

export const loadPath = path => ({
  type: SYNC_GIT_REPO_LOAD_PATH,
  path,
});

export const loadGitCommit = gitCommit => ({
  type: SYNC_GIT_REPO_LOAD_GIT_COMMIT,
  gitCommit,
});

export const loadGitUrl = gitUrl => ({
  type: SYNC_GIT_REPO_LOAD_GIT_URL,
  gitUrl,
});
