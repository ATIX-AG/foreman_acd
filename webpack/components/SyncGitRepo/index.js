import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import './SyncGitRepo.scss';
import SyncGitRepo from './SyncGitRepo';
import * as SyncGitRepoActions from './SyncGitRepoActions';

import {
  selectScmType,
  selectPath,
  selectGitCommit,
  selectGitUrl,
} from './SyncGitRepoSelectors';

const mapStateToProps = state => ({
  scmType: selectScmType(state),
  gitCommit: selectGitCommit(state),
  path: selectPath(state),
  gitUrl: selectGitUrl(state)
});

const mapDispatchToProps = dispatch =>
  bindActionCreators(SyncGitRepoActions, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SyncGitRepo);
