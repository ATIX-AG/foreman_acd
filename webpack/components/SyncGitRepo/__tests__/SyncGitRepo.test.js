import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';

import SyncGitRepo from '../SyncGitRepo';

const noop = () => {};

const fixtures = {
  'should render ansible playbook': {
    data: {
      scmType: "",
      path: "",
      gitCommit: "",
      appDefinitions: "",
      gitUrl: "",
      scmTypes: {},
    },
    initSyncGitRepo: noop,
    loadScmType: noop,
    loadPath: noop,
    loadGitCommit: noop,
    loadGitUrl: noop,
    handleGitRepoSync: noop,
  },
};

describe('SyncGitRepo', () =>
  testComponentSnapshotsWithFixtures(SyncGitRepo, fixtures));
