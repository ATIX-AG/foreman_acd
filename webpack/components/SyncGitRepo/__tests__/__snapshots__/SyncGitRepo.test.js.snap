// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SyncGitRepo should render ansible playbook 1`] = `
<span>
  <div>
    <ScmTypeSelector
      editable={true}
      label="SCM Type *"
      onChange={[Function]}
      options={Object {}}
      selectValue=""
    />
    <div />
    <div
      className="form-group"
    >
      <div
        className="col-md-2"
      />
      <div
        className="col-md-4"
      >
        <pre>
          SCM Type cannot be blank
        </pre>
      </div>
    </div>
  </div>
</span>
`;
